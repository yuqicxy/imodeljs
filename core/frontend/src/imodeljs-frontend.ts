/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/

export * from "./AccuDraw";
export * from "./AccuSnap";
export * from "./AuxCoordSys";
export * from "./CategorySelectorState";
export * from "./ContextRealityModelState";
export * from "./DevTools";
export * from "./DisplayStyleState";
export * from "./ElementLocateManager";
export * from "./EmphasizeElements";
export * from "./EntityState";
export * from "./FrontendLoggerCategory";
export * from "./FrontendRequestContext";
export * from "./FuzzySearch";
export * from "./GeoServices";
export * from "./HitDetail";
export * from "./IModelApp";
export * from "./IModelConnection";
export * from "./ImageUtil";
export * from "./Marker";
export * from "./ModelSelectorState";
export * from "./ModelState";
export * from "./NoRenderApp";
export * from "./NotificationManager";
export * from "./QuantityFormatter";
export * from "./RelativePosition";
export * from "./RenderCompatibility";
export * from "./RenderScheduleState";
export * from "./SelectionSet";
export * from "./Sheet";
export * from "./SolarCalculate";
export * from "./SpatialClassification";
export * from "./Sprites";
export * from "./StandardView";
export * from "./SubCategoriesCache";
export * from "./TentativePoint";
export * from "./TerrainProvider";
export * from "./ViewContext";
export * from "./ViewManager";
export * from "./ViewRect";
export * from "./ViewState";
export * from "./Viewport";
export * from "./oidc/OidcBrowserClient";
export * from "./oidc/OidcDesktopClientRenderer";
export * from "./plugin/Plugin";
export * from "./properties/AngleDescription";
export * from "./properties/BaseQuantityDescription";
export * from "./properties/Description";
export * from "./properties/EditorParams";
export * from "./properties/LengthDescription";
export * from "./properties/PrimitiveTypes";
export * from "./properties/Record";
export * from "./properties/ToolSettingsValue";
export * from "./properties/Value";
export * from "./render/FeatureSymbology";
export * from "./FeatureToggleClient";
export * from "./render/GraphicBuilder";
export * from "./render/MockRender";
export * from "./render/System";
export * from "./render/webgl/PerformanceMetrics";
export * from "./render/webgl/Target";
export * from "./tile/A3xTile";
export * from "./tile/A3xTileIO";
export * from "./tile/BackgroundTerrainTileTree";
export * from "./tile/BingElevation";
export * from "./tile/CesiumWorldTerrainTileTree";
export * from "./tile/GltfReader";
export * from "./tile/IModelTileLoader";
export * from "./tile/ImdlReader";
export * from "./tile/Tile";
export * from "./tile/TileAdmin";
export * from "./tile/TileRequest";
export * from "./tile/TileTree";
export * from "./tile/WebMapTileTree";
export * from "./tools/AccuDrawTool";
export * from "./tools/ClipViewTool";
export * from "./tools/EditManipulator";
export * from "./tools/ElementSetTool";
export * from "./tools/EventController";
export * from "./tools/IdleTool";
export * from "./tools/MeasureTool";
export * from "./tools/PrimitiveTool";
export * from "./tools/SelectTool";
export * from "./tools/Tool";
export * from "./tools/ToolSettings";
export * from "./tools/ToolAdmin";
export * from "./tools/ToolAssistance";
export * from "./tools/ViewTool";

// Set the version number so it can be found at runtime. BUILD_SEMVER is replaced at build time by the webpack DefinePlugin.
declare var BUILD_SEMVER: string;
declare var BUILD_TYPE: string;
if ((typeof (BUILD_SEMVER) !== "undefined") && (typeof window !== "undefined") && window) {
  if (!(window as any).iModelJsVersions)
    (window as any).iModelJsVersions = new Map<string, string>();
  (window as any).iModelJsVersions.set("imodeljs-frontend", BUILD_SEMVER);
  (window as any).iModelJsVersions.set("buildType", BUILD_TYPE);
}

/** @docs-package-description
 * The imodeljs-frontend package always runs in a web browser. It contains classes for [querying iModels and showing views]($docs/learning/frontend/index.md).
 */

/**
 * @docs-group-description IModelApp
 * Classes for configuring and administering an iModel.js application.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description IModelConnection
 * Classes for working with a connection to an [iModel briefcase]($docs/learning/IModels.md)
 */
/**
 * @docs-group-description ElementState
 * Classes for working with the *state* of Elements in the frontend.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description ModelState
 * Classes for working with the *state* of Models in the frontend.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description Tools
 * Classes for [working with Tools]($docs/learning/frontend/Tools.md)
 */
/**
 * @docs-group-description Views
 * Classes for [working with Views]($docs/learning/frontend/Views.md)
 */
/**
 * @docs-group-description LocatingElements
 * Classes for locating and snapping to elements in views.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description AccuDraw
 * AccuDraw provides helpful assistance for creating and modifying elements in a view.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description Notifications
 * Notifications provide feedback to the user of something of interest.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description Plugins
 * Classes for creating and managing runtime [Plugins]($docs/learning/frontend/Plugins.md)
 */
/**
 * @docs-group-description Properties
 * Classes for working with property records and descriptions.
 */
/**
 * @docs-group-description Rendering
 * Classes for rendering the contents of views.
 */
/**
 * @docs-group-description SelectionSet
 * Classes for working with the set of selected elements.
 * See [the learning articles]($docs/learning/frontend/index.md).
 */
/**
 * @docs-group-description OIDC
 * Classes for working with the OpenID Connect (OIDC) protocol
 */
/**
 * @docs-group-description Utils
 * Miscellaneous utility classes.
 */
/**
 * @docs-group-description Logging
 * Logger categories used by this package
 */
/**
 * @docs-group-description SpatialClassification
 * Classes for spatial classification.
 */
/**
 * @docs-group-description Features
 * Classes for Feature Tracking classification.
 */
