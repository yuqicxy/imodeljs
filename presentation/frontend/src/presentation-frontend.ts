/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
/**
 * @module Core
 *
 * @docs-group-description Core
 * Common types used for retrieving presentation data from iModels.
 */
export { PersistenceHelper } from "./PersistenceHelper";
export { Presentation } from "./Presentation";
export { PresentationManager, PresentationManagerProps } from "./PresentationManager";
export { RulesetManager } from "./RulesetManager";
export { RulesetVariablesManager } from "./RulesetVariablesManager";
export { FavoritePropertiesManager, FavoriteProperties } from "./favorite-properties/FavoritePropertiesManager";
export { IFavoritePropertiesStorage, IModelAppFavoritePropertiesStorage } from "./favorite-properties/FavoritePropertiesStorage";

/**
 * @module UnifiedSelection
 *
 * @docs-group-description UnifiedSelection
 * Types related to [unified selection]($docs/learning/presentation/Unified-Selection/index.md).
 */
export { SelectionChangeEvent, SelectionChangeEventArgs, SelectionChangeType, SelectionChangesListener } from "./selection/SelectionChangeEvent";
export { ISelectionProvider } from "./selection/ISelectionProvider";
export { SelectionManager, SelectionManagerProps } from "./selection/SelectionManager";
export { SelectionScopesManager, SelectionScopesManagerProps, getScopeId } from "./selection/SelectionScopesManager";
export { SelectionHandler } from "./selection/SelectionHandler";
export { HiliteSet, HiliteSetProvider } from "./selection/HiliteSetProvider";
export { SelectionHelper } from "./selection/SelectionHelper";

// Set the version number so it can be found at runtime. BUILD_SEMVER is replaced at build time by the webpack DefinePlugin.
declare var BUILD_SEMVER: string;
/* istanbul ignore next */
if ((typeof (BUILD_SEMVER) !== "undefined") && (typeof window !== "undefined") && window) {
  if (!(window as any).iModelJsVersions)
    (window as any).iModelJsVersions = new Map<string, string>();
  (window as any).iModelJsVersions.set("presentation-frontend", BUILD_SEMVER);
}
