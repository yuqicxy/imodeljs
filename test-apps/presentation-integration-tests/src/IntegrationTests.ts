/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
// tslint:disable:no-direct-imports
import * as fs from "fs";
import * as path from "path";
import * as cpx from "cpx";
import "@bentley/presentation-frontend/lib/test/_helpers/MockFrontendEnvironment";
import { I18NOptions } from "@bentley/imodeljs-i18n";
import { Logger, LogLevel } from "@bentley/bentleyjs-core";
import { LoggingNamespaces, RequestPriority } from "@bentley/presentation-common";
import { PresentationProps as PresentationBackendProps } from "@bentley/presentation-backend";
import { PresentationManagerProps as PresentationFrontendProps } from "@bentley/presentation-frontend";
import { NoRenderApp, IModelAppOptions } from "@bentley/imodeljs-frontend";
import { initializeAsync as initializeTesting, terminate as terminateTesting, PresentationTestingInitProps } from "@bentley/presentation-testing";
import { IModelJsConfig } from "@bentley/config-loader/lib/IModelJsConfig";

process.env.NODE_ENV = "development";

IModelJsConfig.init(true);

const copyBentleyBackendAssets = (outputDir: string) => {
  const bentleyPackagesPath = "node_modules/@bentley";
  fs.readdirSync(bentleyPackagesPath).map((packageName) => {
    const packagePath = path.resolve(bentleyPackagesPath, packageName);
    return path.join(packagePath, "lib", "assets");
  }).filter((assetsPath) => {
    return fs.existsSync(assetsPath);
  }).forEach((src) => {
    cpx.copySync(`${src}/**/*`, outputDir);
  });
};

const copyBentleyFrontendAssets = (outputDir: string) => {
  const bentleyPackagesPath = "node_modules/@bentley";
  fs.readdirSync(bentleyPackagesPath).map((packageName) => {
    const packagePath = path.resolve(bentleyPackagesPath, packageName);
    return path.join(packagePath, "lib", "public");
  }).filter((assetsPath) => {
    return fs.existsSync(assetsPath);
  }).forEach((src) => {
    cpx.copySync(`${src}/**/*`, outputDir);
  });
};

class IntegrationTestsApp extends NoRenderApp {
  protected static supplyI18NOptions(): I18NOptions {
    const urlTemplate = "file://" + path.join(path.resolve("lib/public/locales"), "{{lng}}/{{ns}}.json").replace(/\\/g, "/");
    return { urlTemplate };
  }

  public static startup(opts?: IModelAppOptions) {
    NoRenderApp.startup({ ...opts, i18n: this.supplyI18NOptions() });
    cpx.copySync(`assets/**/*`, "lib/assets");
    copyBentleyBackendAssets("lib/assets");
    copyBentleyFrontendAssets("lib/public");
  }
}

export const initialize = async (backendTimeout: number = 0) => {
  // init logging
  Logger.initializeToConsole();
  Logger.setLevelDefault(LogLevel.Error);
  Logger.setLevel(LoggingNamespaces.ECObjects_ECExpressions, LogLevel.Warning);
  Logger.setLevel(LoggingNamespaces.ECPresentation, LogLevel.Warning);

  const backendInitProps: PresentationBackendProps = {
    requestTimeout: backendTimeout,
    rulesetDirectories: ["lib/assets/rulesets"],
    localeDirectories: ["lib/assets/locales"],
    activeLocale: "en-PSEUDO",
    taskAllocationsMap: {
      [RequestPriority.Max]: 1,
    },
  };
  const frontendInitProps: PresentationFrontendProps = {
    activeLocale: "en-PSEUDO",
  };

  const presentationTestingInitProps: PresentationTestingInitProps = {
    backendProps: backendInitProps,
    frontendProps: frontendInitProps,
    frontendApp: IntegrationTestsApp,
    useClientServices: true,
  };

  await initializeTesting(presentationTestingInitProps);
};

export const terminate = () => {
  terminateTesting();
};
