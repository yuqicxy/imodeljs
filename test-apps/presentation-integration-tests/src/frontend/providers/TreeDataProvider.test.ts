/*---------------------------------------------------------------------------------------------
* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
* See LICENSE.md in the project root for license terms and full copyright notice.
*--------------------------------------------------------------------------------------------*/
import { expect } from "chai";
import * as sinon from "sinon";
import { initialize, terminate } from "../../IntegrationTests";
import { IModelConnection } from "@bentley/imodeljs-frontend";
import { PresentationTreeDataProvider } from "@bentley/presentation-components";
import { Presentation } from "@bentley/presentation-frontend";

describe("TreeDataProvider", async () => {

  let imodel: IModelConnection;
  let provider: PresentationTreeDataProvider;

  before(async () => {
    await initialize();
    const testIModelName: string = "assets/datasets/Properties_60InstancesWithUrl2.ibim";
    imodel = await IModelConnection.openSnapshot(testIModelName);
    expect(imodel).is.not.null;
    provider = new PresentationTreeDataProvider(imodel, "SimpleHierarchy");
  });

  after(async () => {
    await imodel.closeSnapshot();
    terminate();
  });

  it("returns root nodes count", async () => {
    const count = await provider.getNodesCount();
    expect(count).to.eq(1);
  });

  it("returns root nodes", async () => {
    const nodes = await provider.getNodes();
    expect(nodes).to.matchSnapshot();
  });

  it("returns root nodes with paging", async () => {
    provider.pagingSize = 5;
    const nodes = await provider.getNodes(undefined, { start: 0, size: 5 });
    expect(nodes.length).to.eq(1);
    expect(nodes).to.matchSnapshot();
  });

  it("does not return root nodes with invalid paging", async () => {
    provider.pagingSize = 5;
    const nodes = await provider.getNodes(undefined, { start: 1, size: 5 });
    expect(nodes.length).to.eq(0);
  });

  it("returns child nodes count", async () => {
    const rootNodes = await provider.getNodes();
    const count = await provider.getNodesCount(rootNodes[0]);
    expect(count).to.eq(1);
  });

  it("returns child nodes", async () => {
    const rootNodes = await provider.getNodes();
    const childNodes = await provider.getNodes(rootNodes[0]);
    expect(childNodes).to.matchSnapshot();
  });

  it("returns child nodes with paging", async () => {
    const rootNodes = await provider.getNodes();
    provider.pagingSize = 5;
    const nodes = await provider.getNodes(rootNodes[0], { start: 0, size: 5 });
    expect(nodes.length).to.eq(1);
    expect(nodes).to.matchSnapshot();
  });

  it("does not return child nodes with invalid paging", async () => {
    const rootNodes = await provider.getNodes();
    provider.pagingSize = 5;
    const nodes = await provider.getNodes(rootNodes[0], { start: 1, size: 5 });
    expect(nodes.length).to.eq(0);
  });

  it("requests backend only once to get first page", async () => {
    const getNodesSpy = sinon.spy(Presentation.presentation.rpcRequestsHandler, "getNodesAndCount");
    provider.pagingSize = 10;

    // request count and first page
    const count = await provider.getNodesCount();
    const nodes = await provider.getNodes(undefined, { start: 0, size: 10 });

    expect(count).to.not.eq(0);
    expect(nodes).to.not.be.undefined;
    expect(getNodesSpy.calledOnce).to.be.true;
  });

});
