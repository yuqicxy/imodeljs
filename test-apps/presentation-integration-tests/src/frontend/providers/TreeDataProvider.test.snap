// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TreeDataProvider returns child nodes 1`] = `
Array [
  Object {
    "__presentation-components/key": Object {
      "pathFromRoot": Array [
        "3da8ffb01a263e88441398a167bf117c",
        "2033788ebe3f7658c6897097d06760fe",
      ],
      "type": "child",
    },
    "description": "child description",
    "icon": "child image id",
    "id": "2033788ebe3f7658c6897097d06760fe/3da8ffb01a263e88441398a167bf117c",
    "label": PropertyRecord {
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "child label",
        "value": "child label",
        "valueFormat": 0,
      },
    },
    "parentId": "3da8ffb01a263e88441398a167bf117c",
  },
]
`;

exports[`TreeDataProvider returns child nodes with paging 1`] = `
Array [
  Object {
    "__presentation-components/key": Object {
      "pathFromRoot": Array [
        "3da8ffb01a263e88441398a167bf117c",
        "2033788ebe3f7658c6897097d06760fe",
      ],
      "type": "child",
    },
    "description": "child description",
    "icon": "child image id",
    "id": "2033788ebe3f7658c6897097d06760fe/3da8ffb01a263e88441398a167bf117c",
    "label": PropertyRecord {
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "child label",
        "value": "child label",
        "valueFormat": 0,
      },
    },
    "parentId": "3da8ffb01a263e88441398a167bf117c",
  },
]
`;

exports[`TreeDataProvider returns root nodes 1`] = `
Array [
  Object {
    "__presentation-components/key": Object {
      "pathFromRoot": Array [
        "3da8ffb01a263e88441398a167bf117c",
      ],
      "type": "root",
    },
    "checkBoxState": 1,
    "description": "root description",
    "hasChildren": true,
    "icon": "root image id",
    "id": "3da8ffb01a263e88441398a167bf117c",
    "isCheckboxDisabled": true,
    "isCheckboxVisible": true,
    "label": PropertyRecord {
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "root label",
        "value": "root label",
        "valueFormat": 0,
      },
    },
    "style": Object {
      "colorOverrides": Object {
        "backgroundColor": 8388863,
        "color": 4278190335,
      },
      "isBold": true,
      "isItalic": true,
    },
  },
]
`;

exports[`TreeDataProvider returns root nodes with paging 1`] = `
Array [
  Object {
    "__presentation-components/key": Object {
      "pathFromRoot": Array [
        "3da8ffb01a263e88441398a167bf117c",
      ],
      "type": "root",
    },
    "checkBoxState": 1,
    "description": "root description",
    "hasChildren": true,
    "icon": "root image id",
    "id": "3da8ffb01a263e88441398a167bf117c",
    "isCheckboxDisabled": true,
    "isCheckboxVisible": true,
    "label": PropertyRecord {
      "property": Object {
        "displayLabel": "Label",
        "name": "node_label",
        "typename": "string",
      },
      "value": Object {
        "displayValue": "root label",
        "value": "root label",
        "valueFormat": 0,
      },
    },
    "style": Object {
      "colorOverrides": Object {
        "backgroundColor": 8388863,
        "color": 4278190335,
      },
      "isBold": true,
      "isItalic": true,
    },
  },
]
`;
